### PR TITLE
from/to quadrature functions

### PR DESCRIPTION
**Pull request by sylviemonet**
**Context:** Add the ABC triples to do the representation transform on a single out_ket wire: from_quad and to_quad.

**Description of the Change:** Two new ABC triples function. It can be used in states, unitaries, and channels directly. 

The higher-level functions must recognize which wire will transform and pick the correct triples for different wires. For example, to transform a single-mode mixed state into quadrature, two of these triples will be called, one is for the out_ket wire of the state, and another one is the conjugate of this triple to transform the out_bra wire of the state.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: mrmustard/physics/triples.py [hunk 1]
```diff
@@ -447,3 +447,42 @@ def fock_damping_Abc(n_modes: int) -> Union[Matrix, Vector, Scalar]:
     c = 1.0 + 0j
 
     return A, b, c
+
+
+def from_quadrature_map_on_single_out_ket_wire_Abc() -> Union[Matrix, Vector, Scalar]:
+    r"""
+    The ``(A, b, c)`` triple of a single out_ket wire map from quadrature representation with ABC Ansatz form into Bargmann representation with ABC Ansatz.
+    Given the real variable in the quadrature representaion of the out_ket wire is :math:`p` and the complex variables for this out_ket wire is :math:`(z^*)`.
+    the indices of the final triple correspond to the variables :math:`(p, z^*)` of the map.
```
Comment (ziofil) [inline mrmustard/physics/triples.py:456]:
```suggestion
    The indices of the final triple correspond to the variables :math:`(p, z^*)` of the map.
```

File: mrmustard/physics/triples.py [hunk 2]
```diff
@@ -447,3 +447,42 @@ def fock_damping_Abc(n_modes: int) -> Union[Matrix, Vector, Scalar]:
     c = 1.0 + 0j
 
     return A, b, c
+
+
+def from_quadrature_map_on_single_out_ket_wire_Abc() -> Union[Matrix, Vector, Scalar]:
```
Comment (ziofil) [inline mrmustard/physics/triples.py:452]:
```suggestion
def from_quadrature_to_out_ket_Abc() -> Union[Matrix, Vector, Scalar]:
```

**Review by ziofil — COMMENTED** [review]
@sylviemonet I think we should add these calculations so that they appear in the documentation. Even I am struggling to follow while I'm reviewing 😅 Feel free to copy/paste the latex formulas if you have them already. @SamFerracin, could you help Yuan with making sure it renders where it should, the page exists etc?

**Comment by SamFerracin** [comment]
@sylviemonet let's work on the docs together when you're back :)

File: mrmustard/physics/triples.py (hunk 2 above)
Comment (SamFerracin) [inline mrmustard/physics/triples.py:452]: from quadrrature abc

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -1,10 +1,10 @@
 # Current develop
 
 ### New features
-* Added one new private CircuitComponent ``_DsMap`` to convert the representations. 
-  [(#385)](https://github.com/XanaduAI/MrMustard/pull/385)
+* Added one new private CircuitComponent ``_BtoQMap`` to convert the representations.
```
Comment (SamFerracin) [inline .github/CHANGELOG.md:4]: remove

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -1,10 +1,10 @@
 # Current develop
 
 ### New features
-* Added one new private CircuitComponent ``_DsMap`` to convert the representations. 
-  [(#385)](https://github.com/XanaduAI/MrMustard/pull/385)
+* Added one new private CircuitComponent ``_BtoQMap`` to convert the representations. 
+  [(#373)](https://github.com/XanaduAI/MrMustard/pull/373)
 
-* Added the function `phase_space(s)` for `State` class to take out the s-parametrized phase space representations' variables covariance matrix and means vector of the state. 
+* Added one new private CircuitComponent ``_DsMap`` to convert the representations.
```
Comment (SamFerracin) [inline .github/CHANGELOG.md:7]: remove

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 5]
```diff
@@ -97,3 +96,30 @@ def __init__(
     @property
     def representation(self) -> Bargmann:
         return Bargmann(*triples.displacement_map_s_parametrized_Abc(self.s, len(self.modes)))
+
+
+class _BtoQMap(CircuitComponent):
+    r"""The change of representation kernel from Bargmann into Quadrature as a ``Unitary``.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:102]: Can we remove the underscores from here and Dsmap? I don't think there's a need to keep them private, since they're in this file which is very clear about containing unphysical maps

Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:102]:
```suggestion
    r"""The kernel for the change of representation from ``Bargmann`` into quadrature.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 6]
```diff
@@ -97,3 +96,30 @@ def __init__(
...
+
+    Used internally as a ``Unitary`` for transformations between representations on the ``Ket`` Wire.
+
+    The ``adjoint`` of this ``CircuitComponent`` denotes the change of representation kernel from Bargmann into Quadrature on the `Bra` Wire.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:106]:
```suggestion
    The ``adjoint`` of this ``CircuitComponent`` denotes the change of representation kernel from ``Bargmann`` into quadrature on the `bra` Wire.
```
Quadrature is not a class, so I'd leave it in small case everywhere (so, do a CTRL+F)

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 7]
```diff
@@ -97,3 +96,30 @@ def __init__(
...
+    The ``dual`` of this ``CircuitComponent`` denotes the change of representation kernel from Quadrature into Bargmann.
+
+    Args:
+        num_modes: The number of modes of this channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:110]:
```suggestion
        modes: The modes of this channel.
```

File: mrmustard/lab_dev/states/base.py [hunk 8]
```diff
@@ -689,6 +718,41 @@ def from_phase_space(
...
+                            full_order_list[len(modes) : 2 * len(modes)],
+                            full_order_list[3 * len(modes) :],
+                        ],
+                        axis=0,
+                    )
+                ),
+            ),
+            idx=list(full_order_list[: 2 * len(modes)]),
+        )
+        ret = DM(name, modes)
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:752]: use `DM._from_attributes`

File: mrmustard/lab_dev/states/base.py [hunk 9]
```diff
@@ -689,6 +718,41 @@ def from_phase_space(
...
+        triple: tuple[ComplexMatrix, ComplexVector, complex],
+        name: Optional[str] = None,
+    ) -> DM:
+        QtoBMap_CC = _BtoQMap(modes).dual.adjoint @ _BtoQMap(modes).dual
+        QtoBMap_A, QtoBMap_b, QtoBMap_c = (
+            QtoBMap_CC.representation.A[0],
+            QtoBMap_CC.representation.b[0],
+            QtoBMap_CC.representation.c[0],
+        )
+        full_order_list = np.arange(4 * len(modes))
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:734]:
```suggestion
        full_order_list = math.arange(4 * len(modes))
```

File: mrmustard/lab_dev/states/base.py [hunk 10]
```diff
@@ -689,6 +718,41 @@ def from_phase_space(
...
         return ret
 
+    @classmethod
+    def from_quadrature(
+        cls,
+        modes: Sequence[int],
+        triple: tuple[ComplexMatrix, ComplexVector, complex],
+        name: Optional[str] = None,
+    ) -> DM:
+        QtoBMap_CC = _BtoQMap(modes).dual.adjoint @ _BtoQMap(modes).dual
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:728]: The logic in this func is very hard to follow. I would add inline comments to explain to posterity what is going on

File: mrmustard/lab_dev/states/base.py [hunk 11]
```diff
@@ -881,6 +945,33 @@ def from_phase_space(
...
+            triple,
+            (QtoBMap_A, QtoBMap_b, QtoBMap_c),
+            idx1=list(np.arange(len(modes))),
+            idx2=list(np.arange(len(modes), 2 * len(modes))),
+        )
+        bargmann_A, bargmann_b, bargmann_c = real_gaussian_integral(
+            (joinedA, joinedb, joinedc),
+            idx=list(np.arange(len(modes))),
+        )
+        ret = Ket(name, modes)
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:971]: `Ket._from_attributes`

File: mrmustard/physics/ansatze.py [hunk 12]
```diff
@@ -557,9 +557,9 @@ def bargmann_Abc_to_phasespace_cov_means(
         The covariance matric, mean vector and coefficient of the state in phase space.
     """
     num_modes = A.shape[-1] // 2
-    Omega = math.J(num_modes).T
-    W = math.conj(math.rotmat(num_modes)).T
+    Omega = np.transpose(math.J(num_modes))
```
Comment (SamFerracin) [inline mrmustard/physics/ansatze.py:560]: use math instead of np

File: mrmustard/physics/ansatze.py [hunk 13]
```diff
@@ -557,9 +557,14 @@ def bargmann_Abc_to_phasespace_cov_means(
         The covariance matric, mean vector and coefficient of the state in phase space.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:557]:
```suggestion
        The covariance matrix, mean vector and coefficient of the state in phase space.
```

File: mrmustard/physics/gaussian_integrals.py [hunk 14]
```diff
@@ -171,6 +171,75 @@ def join_Abc(
...
 
 
+def join_Abc_real(
+    Abc1: Tuple[ComplexMatrix, ComplexVector, complex],
+    Abc2: Tuple[ComplexMatrix, ComplexVector, complex],
+    idx1: Sequence[int],
+    idx2: Sequence[int],
+):
+    r"""Joins two ``(A,b,c)`` triples into a single ``(A,b,c)`` triple by fusing the same meaning indices (idx1 and idx2 give the information about this) in both ``A``
+    matrices and ``b`` vectors.
```
Comment (ziofil) [inline mrmustard/physics/gaussian_integrals.py:181]:
```suggestion
    r"""Direct sum of two ``(A,b,c)`` triples into a single ``(A,b,c)`` triple, where indices corresponding to the same variable are "fused together", by considering their Bargmann function has having the same variables. For example ``idx1=(0,1,2)`` and ``idx2=(1,2,3)`` means that indices 1 and 2 will be fused because they are present on both tuples. This is useful for computing real Gaussian integrals where the variable on either object is the same, rather than a pair of conjugate variables for complex Gaussian integrals.
```

**Review by ziofil — APPROVED** [review]
great!
